### DISCUSSION TITLE
Update multiple rows using Prisma without manual loops

### DISCUSSION THREAD

**Discussion by RajibTheKing** [prisma/prisma#12389] [Q&A]
I have following prisma.schema:

   ```prisma
 model Tag {
      id    Int       @id @default(autoincrement())
      name  String    @unique
      files FileTag[]
    }
    
    model FileTag {
      tag    Tag  @relation(fields: [tagId], references: [id], onDelete: Cascade)
      tagId  Int
      file   File @relation(fields: [fileId], references: [id], onDelete: Cascade)
      fileId Int
    
      @@id([fileId, tagId])
    }

```
Here is my code to update database:

  ```ts
  for (const tagId of tagIds) {
        const tag = await Tags.addFileToTag(parseInt(tagId), fileId);
  };
    
  async addFileToTag(tagId: number, fileId: number) {
      const client = await getDbClient();
  
      return await client.tag.update({
        where: {
          id: tagId,
        },
  
        data: {
          files: {
            create: {
              fileId
            }
          }
        }
      })
    }
```

My goal is reached by this implementation. But the problem is, I don't like this implementation. I am using a loop and calling same **update** query repeatedly. 

Now, I am wondering is there any alternative procedure, (i.e change prisma **update** to **updateMany** query) by removing the loop... that will do the same change on database?

**Comment by austincrim** [comment]
Hey @RajibTheKing 👋 ,

`updateMany` is used to update many rows with the **same** data. In your case, you want to update many rows with **different** data, so it is not suitable in this situation.

If you wanted to improve your code, you could use `Promise.all` to run the queries concurrently and potentially save some time. `Promise.all` expects an array of Promises so we could adjust your code like this:
```ts
const results = await Promise.all(
  tagIds.map(tagId => Tags.addFileToTag(parseInt(tagId), fileId)) // returns an array of Promises
)
```

Let me know if I can clarify anything further!

**Reply by arthurfiorette** [reply]
If he removes the `return await` from the service response, you can change the `Promise.all` to a `Prisma.$transaction`, which may have better performance.

**Reply by petrbela** [reply]
Note that `Promise.all` will assign a connection from the connection pool for each individual operation so if you're doing a lot of updates you might eat up all available connections. `$transaction` is a better solution since all operations will be sent using the same connection.

**Reply by AndrewSouthpaw** [reply]
The tradeoff is that a transaction can lead to heavy contention (depending on the batch size and workload and usage of the existing table), which could have slower speed overall, and potentially even affect your DB performance. So, tread carefully. 😉

**Comment by b05902062** [comment]
there is a related issue #2868

**Comment by mohammedmushtaq3** [comment]
Hey @RajibTheKing 

```
model Tag {
  id    Int       @id @default(autoincrement())
  name  String    @unique
  files FileTag[]
}

model FileTag {
  tag    Tag  @relation(fields: [tagId], references: [id], onDelete: Cascade)
  tagId  Int
  file   File @relation(fields: [fileId], references: [id], onDelete: Cascade)
  fileId Int

  @@id([fileId, tagId])
}

// assumption
model File {
  id      Int       @id @default(autoincrement())
  url     String
  filename     String
  FileTag FileTag[]
}

```
I map through the tagIds which returns prisma update operation  inside a prisma transaction which resolve all operation in one go.

```
  const tagIds = ["1", "2", "4"];
  const fileId = 124;
  const client = await getDbClient();
  const transaction = await client.$transaction(
    tagIds.map((tagId) =>
      client.tag.update({
        where: { id: parseInt(tagId) },
        data: { files: { create: { fileId } } },
      })
    )
  );
  return res.json(transaction);
```
would this be of any help i have tried and tested it and works like a charm.
And do let me know if it worked for you.

**Comment by anton-johansson** [comment]
I brought this up again, in a new discussion (#19765), because I think there's more to discuss. :smile:

**Reply by ryancwalsh** [reply]
https://github.com/prisma/prisma/discussions/19765#discussioncomment-6392390 does look super interesting. Trying it now. Thanks.

**Comment by carlosbaraza** [comment]
[Posted my solution to the problem here](https://github.com/prisma/prisma/discussions/19765#discussioncomment-8244861)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
